{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Section 2: Modeling and effect inference"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "── \u001b[1mAttaching packages\u001b[22m ─────────────────────────────────────── tidyverse 1.3.0 ──\n",
      "\n",
      "\u001b[32m✔\u001b[39m \u001b[34mggplot2\u001b[39m 3.3.2     \u001b[32m✔\u001b[39m \u001b[34mpurrr  \u001b[39m 0.3.4\n",
      "\u001b[32m✔\u001b[39m \u001b[34mtibble \u001b[39m 3.0.4     \u001b[32m✔\u001b[39m \u001b[34mdplyr  \u001b[39m 1.0.2\n",
      "\u001b[32m✔\u001b[39m \u001b[34mtidyr  \u001b[39m 1.1.2     \u001b[32m✔\u001b[39m \u001b[34mstringr\u001b[39m 1.4.0\n",
      "\u001b[32m✔\u001b[39m \u001b[34mreadr  \u001b[39m 1.4.0     \u001b[32m✔\u001b[39m \u001b[34mforcats\u001b[39m 0.5.0\n",
      "\n",
      "── \u001b[1mConflicts\u001b[22m ────────────────────────────────────────── tidyverse_conflicts() ──\n",
      "\u001b[31m✖\u001b[39m \u001b[34mdplyr\u001b[39m::\u001b[32mfilter()\u001b[39m masks \u001b[34mstats\u001b[39m::filter()\n",
      "\u001b[31m✖\u001b[39m \u001b[34mdplyr\u001b[39m::\u001b[32mlag()\u001b[39m    masks \u001b[34mstats\u001b[39m::lag()\n",
      "\n",
      "\n",
      "Attaching package: ‘data.table’\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:dplyr’:\n",
      "\n",
      "    between, first, last\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:purrr’:\n",
      "\n",
      "    transpose\n",
      "\n",
      "\n",
      "Loading required package: S4Vectors\n",
      "\n",
      "Loading required package: stats4\n",
      "\n",
      "Loading required package: BiocGenerics\n",
      "\n",
      "Loading required package: parallel\n",
      "\n",
      "\n",
      "Attaching package: ‘BiocGenerics’\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:parallel’:\n",
      "\n",
      "    clusterApply, clusterApplyLB, clusterCall, clusterEvalQ,\n",
      "    clusterExport, clusterMap, parApply, parCapply, parLapply,\n",
      "    parLapplyLB, parRapply, parSapply, parSapplyLB\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:dplyr’:\n",
      "\n",
      "    combine, intersect, setdiff, union\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:stats’:\n",
      "\n",
      "    IQR, mad, sd, var, xtabs\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:base’:\n",
      "\n",
      "    anyDuplicated, append, as.data.frame, basename, cbind, colnames,\n",
      "    dirname, do.call, duplicated, eval, evalq, Filter, Find, get, grep,\n",
      "    grepl, intersect, is.unsorted, lapply, Map, mapply, match, mget,\n",
      "    order, paste, pmax, pmax.int, pmin, pmin.int, Position, rank,\n",
      "    rbind, Reduce, rownames, sapply, setdiff, sort, table, tapply,\n",
      "    union, unique, unsplit, which, which.max, which.min\n",
      "\n",
      "\n",
      "\n",
      "Attaching package: ‘S4Vectors’\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:data.table’:\n",
      "\n",
      "    first, second\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:dplyr’:\n",
      "\n",
      "    first, rename\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:tidyr’:\n",
      "\n",
      "    expand\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:base’:\n",
      "\n",
      "    expand.grid\n",
      "\n",
      "\n",
      "Loading required package: IRanges\n",
      "\n",
      "\n",
      "Attaching package: ‘IRanges’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:data.table’:\n",
      "\n",
      "    shift\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:dplyr’:\n",
      "\n",
      "    collapse, desc, slice\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:purrr’:\n",
      "\n",
      "    reduce\n",
      "\n",
      "\n",
      "Loading required package: GenomicRanges\n",
      "\n",
      "Loading required package: GenomeInfoDb\n",
      "\n",
      "Loading required package: SummarizedExperiment\n",
      "\n",
      "Loading required package: Biobase\n",
      "\n",
      "Welcome to Bioconductor\n",
      "\n",
      "    Vignettes contain introductory material; view with\n",
      "    'browseVignettes()'. To cite Bioconductor, see\n",
      "    'citation(\"Biobase\")', and for packages 'citation(\"pkgname\")'.\n",
      "\n",
      "\n",
      "Loading required package: DelayedArray\n",
      "\n",
      "Loading required package: matrixStats\n",
      "\n",
      "\n",
      "Attaching package: ‘matrixStats’\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:Biobase’:\n",
      "\n",
      "    anyMissing, rowMedians\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:dplyr’:\n",
      "\n",
      "    count\n",
      "\n",
      "\n",
      "Loading required package: BiocParallel\n",
      "\n",
      "\n",
      "Attaching package: ‘DelayedArray’\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:matrixStats’:\n",
      "\n",
      "    colMaxs, colMins, colRanges, rowMaxs, rowMins, rowRanges\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:purrr’:\n",
      "\n",
      "    simplify\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:base’:\n",
      "\n",
      "    aperm, apply, rowsum\n",
      "\n",
      "\n"
     ]
    }
   ],
   "source": [
    "library(tidyverse)\n",
    "library(data.table)\n",
    "library(DESeq2)\n",
    "library(patchwork)\n",
    "library(ggcorrplot)\n",
    "library(ggpubr)\n",
    "\n",
    "setwd(\"~/work/mpra-v2\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "oligoCountPath = 'data/1KG_bartenderOligoSums_novaSeq.txt'\n",
    "\n",
    "oligoCounts = fread(oligoCountPath)\n",
    "oligoCounts = oligoCounts[,c(\"Pool\",\"VarID\",\"geneID\",\n",
    "                         \"chrom\",\"pos\",\"strand\",\n",
    "                         \"ref\",\"alt\",\"haploFlag\",\n",
    "                         \"cDNA1_ref\",\"cDNA2_ref\",\"cDNA3_ref\",\n",
    "                         \"plasmid1_ref\",\"plasmid2_ref\",\"plasmid3_ref\",\n",
    "                         \"cDNA1_alt\",\"cDNA2_alt\",\"cDNA3_alt\",\n",
    "                         \"plasmid1_alt\",\"plasmid2_alt\",\"plasmid3_alt\")]\n",
    "oligoCounts = oligoCounts[which(rowMeans(oligoCounts[,10:21]) > 150),]\n",
    "oligoCounts = subset(oligoCounts, haploFlag == \"Base\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "design = data.frame(\"material\" = rep(c(rep(\"RNA\",3),rep(\"DNA\",3)),2), \"allele\" = c(rep(\"ref\",6),rep(\"alt\",6)))\n",
    "design$replicate = rep(c(\"cDNA1\",\"cDNA2\",\"cDNA3\",\"plasmid1\",\"plasamid2\",\"plasmid3\"),2)\n",
    "rownames(design) = names(oligoCounts)[10:21]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Part 1: DESeq2 with interaction term"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Build and execute DESeq2 call and retrieve summary statistics\n",
    "deseq_object = DESeqDataSetFromMatrix(countData=oligoCounts[,10:21],\n",
    "                                      colData = design,\n",
    "                                      design = ~material + allele + material:allele)\n",
    "deseq_object = DESeq(deseq_object)\n",
    "\n",
    "# Format separate contrasts and bind together\n",
    "results_material = results(deseq_object, contrast = c(\"material\",\"RNA\",\"DNA\"))\n",
    "results_oligo = results(deseq_object, contrast = c(\"allele\",\"ref\",\"alt\"))\n",
    "results_allele = results(deseq_object)\n",
    "\n",
    "names(results_material) = paste0(names(results_material),\"_\", \"expr\")\n",
    "names(results_oligo) = paste0(names(results_oligo),\"_\", \"oligo\")\n",
    "names(results_allele) = paste0(names(results_allele),\"_\", \"allele\")\n",
    "\n",
    "results_DESeq2 = cbind(oligoCounts, as.data.frame(results_material), as.data.frame(results_oligo), as.data.frame(results_allele))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "write.table(results_DESeq2, \"sumstats/1KG_novaSeq_DESeq2_interaction.txt\", sep = \"\\t\", quote = F, row.names = F, col.names = T)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Part 2: DESeq2 as formulated by MLove (http://rstudio-pubs-static.s3.amazonaws.com/275642_e9d578fe1f7a404aad0553f52236c0a4.html) with an interaction term"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "design_love = data.frame(\"material\" = factor(rep(c(rep(\"RNA\",3),rep(\"DNA\",3)),2)),\n",
    "                    \"allele\" = factor(c(rep(\"ref\",6),rep(\"alt\",6))),\n",
    "                    \"sample\" = factor(rep(c(\"cDNA1\",\"cDNA2\",\"cDNA3\",\"plasmid1\",\"plasmid2\",\"plasmid3\"),2)))\n",
    "rownames(design_love) = names(oligoCounts)[10:21]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "estimating size factors\n",
      "\n",
      "estimating dispersions\n",
      "\n",
      "gene-wise dispersion estimates\n",
      "\n",
      "mean-dispersion relationship\n",
      "\n",
      "-- note: fitType='parametric', but the dispersion trend was not well captured by the\n",
      "   function: y = a/x + b, and a local regression fit was automatically substituted.\n",
      "   specify fitType='local' or 'mean' to avoid this message next time.\n",
      "\n",
      "final dispersion estimates\n",
      "\n",
      "fitting model and testing\n",
      "\n"
     ]
    }
   ],
   "source": [
    "# Build and execute DESeq2 call and retrieve summary statistics\n",
    "deseq_object_love = DESeqDataSetFromMatrix(countData=oligoCounts[,10:21],\n",
    "                                           colData = design_love, \n",
    "                                           design = ~sample + allele)\n",
    "deseq_object_love$sample.n = factor(rep(c(\"A\",\"B\",\"C\"),4))\n",
    "design(deseq_object_love) =  ~material + material:sample.n + material:allele\n",
    "deseq_object_love = DESeq(deseq_object_love)\n",
    "\n",
    "# Generate and bind results\n",
    "#results_expr_love = results(deseq_object_love, contrast = c('material','RNA','DNA'))\n",
    "results_expr_love = results(deseq_object_love, contrast = c(0,1,-1/3,1/3,-1/3,1/3,-1/2,1/2))\n",
    "results_allele_love = results(deseq_object_love, contrast = list('materialDNA.alleleref','materialRNA.alleleref'))\n",
    "\n",
    "names(results_expr_love) = paste0(names(results_expr_love),\"_\", \"expr\")\n",
    "names(results_allele_love) = paste0(names(results_allele_love),\"_\", \"allele\")\n",
    "\n",
    "results_DESeq2_love = cbind(oligoCounts, \n",
    "                            as.data.frame(results_expr_love),\n",
    "                            as.data.frame(results_allele_love))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "#write.table(results_DESeq2_love, \"sumstats/1KG_novaSeq_DESeq2_Love_Base.txt\", quote = F, sep = '\\t', row.names = F, col.names = T)\n",
    "write.table(results_DESeq2_love, \"sumstats/1KG_novaSeq_DESeq2_Love_Base2.txt\", quote = F, sep = '\\t', row.names = F, col.names = T)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Part 3: DESeq2 using linear contrasts"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "design_contrast = data.frame(\"material\" = factor(rep(c(rep(\"RNA\",3),rep(\"DNA\",3)),2)),\n",
    "                    \"allele\" = factor(c(rep(\"ref\",6),rep(\"alt\",6))),\n",
    "                    \"sample\" = factor(rep(c(\"cDNA1\",\"cDNA2\",\"cDNA3\",\"plasmid1\",\"plasmid2\",\"plasmid3\"),2)),\n",
    "                    \"dnaAllele\" = c(0,0,0,0,0,0,0,0,0,1,1,1),\n",
    "                    \"rnaAllele\" = c(0,0,0,0,0,0,1,1,1,0,0,0))\n",
    "rownames(design_contrast) = names(oligoCounts)[10:21]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "deseq_object_contrast = DESeqDataSetFromMatrix(countData=oligoCounts[,10:21],\n",
    "                                               colData = design_contrast,\n",
    "                                               design = ~0 + sample + dnaAllele + rnaAllele)\n",
    "deseq_object_contrast = DESeq(deseq_object_contrast)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "results_expr_contrast = results(deseq_object_contrast, contrast = list(c(\"samplecDNA1\", \"samplecDNA2\",\"samplecDNA3\"),\n",
    "                                                              c(\"sampleplasmid1\",\"sampleplasmid2\",\"sampleplasmid3\")))\n",
    "results_allele_contrast = results(deseq_object_contrast, contrast = list(\"dnaAllele\",\"rnaAllele\"))\n",
    "\n",
    "names(results_expr_contrast) = paste0(names(results_expr_contrast),\"_\", \"expr\")\n",
    "names(results_allele_contrast) = paste0(names(results_allele_contrast),\"_\", \"allele\")\n",
    "\n",
    "results_DESeq2_contrast = cbind(oligoCounts,\n",
    "                                as.data.frame(results_expr_contrast),\n",
    "                                as.data.frame(results_allele_contrast))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "write.table(results_DESeq2_contrast, \"sumstats/1KG_novaSeq_DESeq2_contrast.txt\", quote = F, sep = '\\t', row.names = F, col.names = T)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "love = fread(\"sumstats/1KG_novaSeq_DESeq2_Love_Base.txt\")\n",
    "love = separate_rows(data = love, sep = \",\",geneID)\n",
    "brca1 = subset(love,geneID == \"ENSG00000012048\")\n",
    "\n",
    "options(repr.plot.width = 18, repr.plot.height = 6)\n",
    "coeff = 0.19\n",
    "ggplot(brca1) + \n",
    "    geom_vline(xintercept = unlist(subset(brca1,padj_expr<5e-2 & padj_allele < 5e-2,select=pos)),\n",
    "               color = \"magenta\", size = 1) +\n",
    "    geom_point(aes(x = pos, y = -log(pvalue_expr),\n",
    "                   size = -log(pvalue_expr),\n",
    "                   color = ifelse(padj_expr < 5e-2,\"red\",\"black\"))) + \n",
    "    geom_point(aes(x = pos, y = -log(pvalue_allele)/coeff,\n",
    "                   size = -log(pvalue_allele)/coeff,\n",
    "                   color = ifelse(padj_allele < 5e-2,\"blue\",\"black\"))) + \n",
    "    theme_pubr(base_size = 20) + \n",
    "    scale_x_continuous(limits = c(41175000,41400000), breaks = c(41200000,41300000,41400000)) + scale_color_identity() +\n",
    "    scale_y_continuous(name = \"-log(expression p-value)\",sec.axis = sec_axis(~.*coeff, name=\"-log(allelic p-value)\")) +\n",
    "    theme(legend.position = \"none\") + xlab(\"chr17 genomic position\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "love = fread(\"sumstats/1KG_novaSeq_DESeq2_Love_Base.txt\")\n",
    "love = separate_rows(data = love, sep = \",\",geneID)\n",
    "brca1 = subset(love,geneID == \"ENSG00000012048\")\n",
    "\n",
    "options(repr.plot.width = 18, repr.plot.height = 6)\n",
    "coeff = 0.19\n",
    "ggplot(brca1) + \n",
    "    geom_point(aes(x = pos, y = -log(pvalue_expr),\n",
    "                   size = -log(pvalue_expr),\n",
    "                   color = ifelse(padj_expr < 5e-2,\"red\",\"black\"))) + \n",
    "    geom_point(aes(x = pos, y = -log(pvalue_allele)/coeff,\n",
    "                   size = -log(pvalue_allele)/coeff,\n",
    "                   color = ifelse(padj_allele < 5e-2,\"blue\",\"black\"))) + \n",
    "    theme_pubr(base_size = 20) + \n",
    "    scale_x_continuous(limits = c(41175000,41400000), breaks = c(41200000,41300000,41400000)) + scale_color_identity() +\n",
    "    scale_y_continuous(name = \"-log(expression p-value)\",sec.axis = sec_axis(~.*coeff, name=\"-log(allelic p-value)\")) +\n",
    "    theme(legend.position = \"none\") + xlab(\"chr17 genomic position\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "R version 3.6.0 (2019-04-26)\n",
       "Platform: x86_64-pc-linux-gnu (64-bit)\n",
       "Running under: CentOS Linux 7 (Core)\n",
       "\n",
       "Matrix products: default\n",
       "BLAS/LAPACK: /usr/lib64/libopenblas-r0.3.3.so\n",
       "\n",
       "locale:\n",
       " [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              \n",
       " [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    \n",
       " [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   \n",
       " [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 \n",
       " [9] LC_ADDRESS=C               LC_TELEPHONE=C            \n",
       "[11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       \n",
       "\n",
       "attached base packages:\n",
       "[1] parallel  stats4    stats     graphics  grDevices utils     datasets \n",
       "[8] methods   base     \n",
       "\n",
       "other attached packages:\n",
       " [1] ggpubr_0.4.0                ggcorrplot_0.1.3           \n",
       " [3] patchwork_1.1.0             DESeq2_1.26.0              \n",
       " [5] SummarizedExperiment_1.16.1 DelayedArray_0.12.3        \n",
       " [7] BiocParallel_1.20.1         matrixStats_0.57.0         \n",
       " [9] Biobase_2.46.0              GenomicRanges_1.38.0       \n",
       "[11] GenomeInfoDb_1.22.1         IRanges_2.20.2             \n",
       "[13] S4Vectors_0.24.4            BiocGenerics_0.32.0        \n",
       "[15] data.table_1.13.2           forcats_0.5.0              \n",
       "[17] stringr_1.4.0               dplyr_1.0.2                \n",
       "[19] purrr_0.3.4                 readr_1.4.0                \n",
       "[21] tidyr_1.1.2                 tibble_3.0.4               \n",
       "[23] ggplot2_3.3.2               tidyverse_1.3.0            \n",
       "\n",
       "loaded via a namespace (and not attached):\n",
       " [1] colorspace_2.0-0       ggsignif_0.6.0         rio_0.5.16            \n",
       " [4] ellipsis_0.3.1         IRdisplay_0.7.0        htmlTable_2.1.0       \n",
       " [7] XVector_0.26.0         base64enc_0.1-3        fs_1.5.0              \n",
       "[10] rstudioapi_0.12        bit64_4.0.5            AnnotationDbi_1.48.0  \n",
       "[13] fansi_0.4.1            lubridate_1.7.9        xml2_1.3.2            \n",
       "[16] splines_3.6.0          geneplotter_1.64.0     knitr_1.30            \n",
       "[19] IRkernel_1.1.1         Formula_1.2-4          jsonlite_1.7.1        \n",
       "[22] broom_0.7.2            annotate_1.64.0        cluster_2.1.0         \n",
       "[25] dbplyr_2.0.0           png_0.1-7              compiler_3.6.0        \n",
       "[28] httr_1.4.2             backports_1.2.0        assertthat_0.2.1      \n",
       "[31] Matrix_1.2-18          cli_2.1.0              htmltools_0.5.0       \n",
       "[34] tools_3.6.0            gtable_0.3.0           glue_1.4.2            \n",
       "[37] GenomeInfoDbData_1.2.2 Rcpp_1.0.5             carData_3.0-4         \n",
       "[40] cellranger_1.1.0       vctrs_0.3.4            xfun_0.19             \n",
       "[43] ps_1.4.0               openxlsx_4.2.3         rvest_0.3.6           \n",
       "[46] lifecycle_0.2.0        rstatix_0.6.0          XML_3.99-0.3          \n",
       "[49] getPass_0.2-2          zlibbioc_1.32.0        scales_1.1.1          \n",
       "[52] hms_0.5.3              RColorBrewer_1.1-2     curl_4.3              \n",
       "[55] memoise_1.1.0          gridExtra_2.3          rpart_4.1-15          \n",
       "[58] latticeExtra_0.6-29    stringi_1.5.3          RSQLite_2.2.1         \n",
       "[61] genefilter_1.68.0      checkmate_2.0.0        zip_2.1.1             \n",
       "[64] repr_1.1.0             rlang_0.4.8            pkgconfig_2.0.3       \n",
       "[67] bitops_1.0-6           evaluate_0.14          lattice_0.20-41       \n",
       "[70] htmlwidgets_1.5.2      bit_4.0.4              tidyselect_1.1.0      \n",
       "[73] magrittr_1.5           R6_2.5.0               generics_0.1.0        \n",
       "[76] Hmisc_4.4-1            pbdZMQ_0.3-3.1         DBI_1.1.0             \n",
       "[79] pillar_1.4.6           haven_2.3.1            foreign_0.8-76        \n",
       "[82] withr_2.3.0            abind_1.4-5            survival_3.2-7        \n",
       "[85] RCurl_1.98-1.2         nnet_7.3-14            car_3.0-10            \n",
       "[88] modelr_0.1.8           crayon_1.3.4           uuid_0.1-4            \n",
       "[91] jpeg_0.1-8.1           locfit_1.5-9.4         grid_3.6.0            \n",
       "[94] readxl_1.3.1           blob_1.2.1             reprex_0.3.0          \n",
       "[97] digest_0.6.27          xtable_1.8-4           munsell_0.5.0         "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "sessionInfo()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R",
   "language": "R",
   "name": "ir"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "3.6.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
